File: 0203_remove linked list elements/Remove Linked List Elements2.java
class Solution {
    public ListNode removeElements(ListNode head, int val) {
        // 设置一个虚拟头结点，这样每个节点都将有前一个节点，这样不用对第一个元素进行特殊考虑了
        ListNode dummyHead = new ListNode(-1);
        dummyHead.next = head;
        ListNode prev = dummyHead;
        while(prev.next != null){
            if(prev.next.val == val)
                prev.next = prev.next.next;
            else
                prev = prev.next;
        }
        return dummyHead.next;
    }
}

File: 0203_remove linked list elements/Remove Linked List Elements2.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
  ListNode* removeElements(ListNode* head, int val) {
    // 递归解法
    if(head==NULL){
      return head;
    }
    ListNode *res = removeElements(head->next,val);
    if(head->val==val){
      return res;
    } else{
      head->next=res;
      return head;
    }
  }
};